connectors/memory/memory.go

```diff
@@ -0,0 +1,493 @@
+	mid := int64(binary.BigEndian.Uint16(u[4:6]))
+	hi := int64((binary.BigEndian.Uint16(u[6:8]) & 0x0fff))
+	return low + (mid << 32) + (hi << 48)
+}
+
+// compareType compares a single DOSA field based on the type. This code assumes the types of each
+// of the columns are the same, or it will panic
+func compareType(d1 dosa.FieldValue, d2 dosa.FieldValue) int8 {
+	switch d1 := d1.(type) {
+	case dosa.UUID:
+		u1 := uuid.FromStringOrNil(string(d1))
+		u2 := uuid.FromStringOrNil(string(d2.(dosa.UUID)))
+		if u1.Version() != u2.Version() {
+			if u1.Version() < u2.Version() {
+				return -1
+			}
+			return 1
+		}
+		if u1.Version() == 1 {
+			// compare time UUIDs
+			t1 := timeFromUUID(u1)
+			t2 := timeFromUUID(u2)
+			if t1 == t2 {
+				return 0
+			}
+			if t1 < t2 {
+				return -1
+			}
+			return 1
+		}
+
+		// version
+		if string(d1) == string(d2.(dosa.UUID)) {
+			return 0
+		}
+		if string(d1) < string(d2.(dosa.UUID)) {
+			return -1
+		}
+		return 1
+	case string:
+		if d1 == d2.(string) {
+			return 0
+		}
+		if d1 < d2.(string) {
+			return -1
+		}
+		return 1
+	case int64:
+		if d1 == d2.(int64) {
+			return 0
+		}
+		if d1 < d2.(int64) {
+			return -1
+		}
+		return 1
+	case int32:
+		if d1 == d2.(int32) {
+			return 0
+		}
+		if d1 < d2.(int32) {
+			return -1
+		}
+		return 1
+	case float64:
+		if d1 == d2.(float64) {
+			return 0
+		}
+		if d1 < d2.(float64) {
+			return -1
+		}
+		return 1
+	case []byte:
+		c := bytes.Compare(d1, d2.([]byte))
+		if c == 0 {
+			return 0
+		}
+		if c < 0 {
+			return -1
+		}
+		return 1
+	case time.Time:
+		if d1.Equal(d2.(time.Time)) {
+			return 0
+		}
+		if d1.Before(d2.(time.Time)) {
+			return -1
+		}
+		return 1
+	case bool:
+		if d1 == d2.(bool) {
+			return 0
+		}
+		if d1 == false {
+			return -1
+		}
+		return 1
+	}
+	panic(d1)
+}
+
+// CreateIfNotExists inserts a row if it isn't already there. The basic flow is:
+// Find the partition, if it's not there, then create it and insert the row there
+// If the partition is there, and there's data in it, and there's no clustering key, then fail
+// Otherwise, search the partition for the exact same clustering keys. If there, fail
+// if not, then insert it at the right spot (sort.Search does most of the heavy lifting here)
+func (c *Connector) CreateIfNotExists(_ context.Context, ei *dosa.EntityInfo, values map[string]dosa.FieldValue) error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	if c.data == nil {
+		c.data = make(map[string]map[string][]map[string]dosa.FieldValue)
+	}
+	if c.data[ei.Def.Name] == nil {
+		c.data[ei.Def.Name] = make(map[string][]map[string]dosa.FieldValue)
+	}
+	entityRef := c.data[ei.Def.Name]
+	encodedPartitionKey := partitionKeyBuilder(ei, values)
+	if entityRef[encodedPartitionKey] == nil {
+		entityRef[encodedPartitionKey] = make([]map[string]dosa.FieldValue, 0, 1)
+	}
+	partitionRef := entityRef[encodedPartitionKey]
+	// no data in this partition? easy out!
+	if len(partitionRef) == 0 {
```

Comment:
I really do want this code to look at len(partitionRef) not check for existence. It might exist but it's an empty map. See this code for more details.

https://play.golang.org/p/Oo8ZtQGIN3